src/v/storage/disk_log_impl.cc

```diff
@@ -2521,6 +2521,8 @@ disk_log_impl::remove_prefix_full_segments(truncate_prefix_config cfg) {
       },
       [this] {
           auto ptr = _segs.front();
+          // mark as tombstone before next scheduling point
```

Comment:
I was thinking about this too but I think the bug here is the segment should be marked tombstone() in the same continuation as _segs.pop_front(), because there is a scheduling point right at invoking of remove_segment_permanently().

So the bug (as I understood) is


fiber0

_segs.pop_front();
<scheduling point> (remove_prefix_full_segments not yet invoked)

fiber 1

do_append() -- lost append due to prefix truncation. 

fiber2

make_appender() -- doesn't see the appended offset from fiber 1 (because the appended segment is already popped .. check offset_stats disk_log_impl::offsets()), this results in a duplicate base offset.

Reply:
My question is more around `remove_segment_permanently()` being called as a future rather than as a coroutine. Since we're returning a future from this method, isn't `_segs.pop_front()` happening in the same task as `s->tombstone()` in `remove_segment_permanently()`? I thought we don't incur the scheduling point if we are just constructing the future.

In any case, I think the race was:

0. log has [0-9] and one segment s1
1. appender a1 writes a batch with offset 10 to s1
2. remove_prefix_full_segment is called, and this removes the segment from the segment list (but not marked tombstoned). this leaves the log empty, and the new start offset is 10.
3. a1 doesn't realize this, and proceeds to write to s1, offset 10 is assigned and returned to Raft
4. appender a2 is constructed with the start offset 10
5. a2 writes to s2, assigning another batch offset 10 and returns it to Raft

With this change I think it's meant to go:

0. log has [0-9] and one segment s1
1. appender a1 with offset 10 writes to s1
2. remove_prefix_full_segment is called, and this removes the segment from the segment list _and_ marks the segment tombstoned. this leaves the log empty, **but also indicates to the appender that s1 is tombstoned**
3. **a1 realizes this and creates a new segment s2 with dirty offset 10**
4. appender a2 is constructed with the s1's dirty offset + 1, offset 11
5. a2 writes to s2, assigning another batch offset 11 and returns it to Raft

IIUC, actually I now I think there's still the following race:

0. log has [0-9] and one segment s1
1. appender a1 with offset 10 writes to s1
2. **a1 proceeds to begin writing to s1, offset 10 is assigned to the in-flight batch. a1 proceeds past the segment_is_appendable check and the read lock is held**
3. remove_prefix_full_segment is called, and this removes the segment from the segment list _and_ marks the segment tombstoned. this leaves the log empty, but **also indicates to the appender that s1 is tombstoned... but....**
4. **a1 continues onto append_batch_to_segment() with offset 10**
5. appender a2 is constructed with the start offset 10
6. a2 writes to s2, assigning another batch offset 10 and returns it to Raft

Reply:
I agree that the problem isn't solved. It almost exactly as Andrew described above in the third example. I think the problem is that the `next_offset` for appender is computed before the segment lock is taken, this is the source of issues. 
I've changed the code to check the lock before deciding if the segment should be deleted rather than checking the log only to defer the segment deletion, this way the prefix truncate loop will yield to appender if needed

Reply:
> My question is more around remove_segment_permanently() being called as a future rather than as a coroutine. Since we're returning a future from this method, isn't _segs.pop_front() happening in the same task as s->tombstone() in remove_segment_permanently()? I thought we don't incur the scheduling point if we are just constructing the future.

yep I think you are right, I got confused by the comment about scheduling point and haven't looked too closely.

> In any case, I think the race was:

yep exactly.


> IIUC, actually I now I think there's still the following race:

that makes sense to me, nice find.